### ISSUE TITLE
🌋 E|F ➾ Proposal to reconcile E2E tests with CICD

### ISSUE THREAD

**Issue by michaelkernaghan** [ecadlabs/taqueria#1555]
### 🚥 Status

- [X] 🔮 Product/Feature Design
- [X] ✍🏻 Architecture/Implementation Planning
- [ ] 🏗️ Under Construction
- [ ] ⛳ Feature Complete
- [ ] ⛵ GTM / Release

### Sub-tickets

- [x] #1600 
- [ ]  #1599 

### ✅ To Do

get agreement
work out POC
coordinate with SRE for CICD

### 📐 Feature Description

Whereas: 

The E2E tests are valuable and should be frequently used, but they can be fragile and require maintenance.
The time needed for the maintenance can interfere with CICD efficiency.

Be it resolved that:

The E2E tests as a whole will be dropped from CICD and instead will be maintained, run, and reported on by SDETs on the Taqueria team working with SREs.
The CICD will run a succinct representative selection of E2E  tests, called an E2E smoke test, that satisfies the following. 

Proof of all origination 
Proof of compilation 
UI Sanity Check, 
One test for each module
Others to be TBD, not to exceed 10 in total

This can be accomplished by adding taqueria-e2e-smoke.spec.ts to the ~/taqueria/tests/e2e/ direcory

The E2E smoke test can be called by

npm run test:e2e:smoke -w tests

The test package.json scripts set would add.

"test:e2e:smoke": "TEST=true jest --config jest.config.e2e.ts --detectOpenHandles --detectLeaks taqueria-e2e-smoke.spec.ts"
or
"	"test:e2e:smoke": "../bin/before-all-tests.sh && TEST=true jest --config jest.config.e2e.ts --maxWorkers=1 --testNamePattern=smoke","

### 🚩 Problem or Use Case

The E2E tests are a comprehensive suite of checks for expected outputs from Taqueria CLI combinations, using stored files as expectations to match. There are several such tests for each module, and the time taken to run the whole suite is enough to dissuade developers from running them before a code commit. Also, tests must be rewritten when anything on an output changes, which is frequent at this stage of development. While waiting to be rewritten, they can prevent CICD operations by failing in the pipeline.

To improve this situation, we need to use the E2E test differently, as described above.

Additional consideration:

The end-to-end test suite has many tests that compare the output from a cli call with a prepared file that shows the expected outcome.
Tests like this have two failure modes. The test may fail because something is wrong with the system that provides an output. Or, the test may fail because something in the output has changed.
In the first case, that should be caught at the unit or integration test level and fixed by the devs before committing.
If the test changes because of an output update, then the test is not giving you any new information. You already know that there were changes to the output. The test catches this change, and the fix for the test is to make it match the outcome again. The problem with these tests is they lack probative value. They don't tell you anything new about the program you didn't already know.
To make the test pass again, someone has to update it to match the expected output. But that is the same as having closed an issue about the change; it is information everyone already has. The test adds nothing to the understanding of the system.
This problem gets worse when these tests are in a CICD. If there is a change to wording in the output, the E2E test will fail, stopping a PR. The CICD has to wait till the test is updated. But the update is an after-the-fact change to reflect something already tracked elsewhere. Even if someone anticipates this and corrects the test before it runs and fails on CICD, it is still, after the fact catching up and not a probative examination.

### ⚖️ Acceptance Criteria

✨ Features: 
- [ ] ... no more E2E test CICD conflicts

### 🔭 Scope

🏎️ In Scope: 
- ... E2E tests
 
 🙅‍♂️ Out of Scope: 
-  unit tests
-  integration tests
 
 

### 🦄 Sizing

3

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by michaelkernaghan** [comment]
I have had to abandon the quest for a succinct version of this E2E suite. The required modularity of the code is not there, and I have discovered extensive mixing of end-user detail with implementation detail. 
This separation can be achieved by rewriting the code in a proper DSL library to encapsulate the implementation details.
This will make the suite more maintainable as it is not subject to breakage due to refactoring of implementations, etc. 
The existing E2E test suite should continue to run but not be extended until it can be replaced. Its scope of tests, and sequences of cli actions, will serve as acceptance requirements for this migration.

Here is a code sample of one such DSL in action (https://github.com/gmrchk/cli-testing-library)
```
import CLITestingLibary from '@gmrchk/cli-testing-library';

import { prepareEnvironment } from '@gmrchk/cli-testing-library';

describe('Basic E2E for taq CLI', () => {
    it('taq --help runs successfully', async () => {
        const { execute, cleanup } = await prepareEnvironment();

        const { code } = await execute(
            'taq',
            ' --help'
        );
        
        expect(code).toBe(0);

        await cleanup();
    });
```
